### PR TITLE
langchain[minor]: Firecrawl Document Loader

### PR DESCRIPTION
**Pull request by nickscamara**
Added the [FireCrawl](https://firecrawl.dev/) document loader. Firecrawl crawls and convert any website into LLM-ready data. It crawls all accessible subpages and give you clean markdown for each.

Similar to https://github.com/langchain-ai/langchain/pull/20364 but JS/TS :)

Thank you!

### CODE REVIEW

File: docs/core_docs/docs/integrations/document_loaders/web_loaders/firecrawl.mdx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+Firecrawl offers 2 modes: `scrape` and `crawl`. In `scrape` mode, Firecrawl will only scrape the page you provide. In `crawl` mode, Firecrawl will crawl the entire website.
+
+
+import CodeBlock from "@theme/CodeBlock";
+import Example from "@examples/document_loaders/firecrawl.ts";
+
+import IntegrationInstallTooltip from "@mdx_components/integration_install_tooltip.mdx";
+
+<IntegrationInstallTooltip></IntegrationInstallTooltip>
```
Comment (bracesproul) [inline docs/core_docs/docs/integrations/document_loaders/web_loaders/firecrawl.mdx:34]: No need for an `IntegrationInstallTooltip` for non LC deps. You only this for langchain package installs

File: docs/core_docs/docs/integrations/document_loaders/web_loaders/firecrawl.mdx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+[FireCrawl](https://firecrawl.dev) crawls and convert any website into LLM-ready data. It crawls all accessible subpages and give you clean markdown and metadata for each. No sitemap required.
+
+FireCrawl handles complex tasks such as reverse proxies, caching, rate limits, and content blocked by JavaScript. Built by the [mendable.ai](https://mendable.ai) team.
+
+This guide shows how to scrap and crawl entire websites and load them using the `FireCrawlLoader` in LangChain.
+
+## Setup
+
+You'll need to sign up and retrieve your [FireCrawl API key](https://firecrawl.dev).
```
Comment (bracesproul) [inline docs/core_docs/docs/integrations/document_loaders/web_loaders/firecrawl.mdx:19]:
```suggestion
You'll need to sign up and retrieve your [FireCrawl API key](https://firecrawl.dev).
```
I would add some mention of the 100 free credits. I could easily see a user reading up until the API key requirement part and getting discouraged before they're even able to try it out

File: examples/src/document_loaders/firecrawl.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { FireCrawlLoader } from "langchain/document_loaders/web/firecrawl";
+
+const loader = new FireCrawlLoader({
+    url: "https://firecrawl.dev", // The URL to scrape
+    apiKey: process.env.FIRECRAWL_API_KEY, //  This will be loaded from your .env too
```
Comment (bracesproul) [inline examples/src/document_loaders/firecrawl.ts:5]:
```suggestion
    apiKey: process.env.FIRECRAWL_API_KEY, // Optional, defaults to `FIRECRAWL_API_KEY` in your env.
```

File: langchain/src/document_loaders/web/firecrawl.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ * Interface representing the parameters for the Firecrawl loader. It
+ * includes properties such as the URL to scrape or crawl and the API key.
+ */
+interface FirecrawlLoaderParameters {
+  /**
+   * URL to scrape or crawl
+   */
+  url: string;
+  apiKey?: string;
+  mode?: "crawl" | "scrape";
```
Comment (bracesproul) [inline langchain/src/document_loaders/web/firecrawl.ts:16]: Let's add JSDoc here documenting what the default values are for both of these

File: langchain/src/document_loaders/web/firecrawl.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  metadata: Record<string, unknown>;
+}
+
+/**
+ * Class representing a document loader for loading data from
+ * Firecrawl (firecrawl.dev). It extends the BaseDocumentLoader class.
+ * @example
+ * ```typescript
+ * const loader = new FireCrawlLoader({ url: "{url}", apiKey: "{apiKey}", mode: "crawl" });
+ * const docs = await loader.load();
```
Comment (bracesproul) [inline langchain/src/document_loaders/web/firecrawl.ts:30]:
```suggestion
 * ```typescript
 * const loader = new FireCrawlLoader({
 *   url: "{url}",
 *   apiKey: "{apiKey}",
 *   mode: "crawl"
 * });
 * const docs = await loader.load();
```
will look better in the api refs if it has formatting

File: langchain/src/document_loaders/web/firecrawl.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    this.mode = mode;
+    this.params = params;
+  }
+
+  /**
+   * Loads the data from the Firecrawl.
+   * @returns An array of Documents representing the retrieved data.
+   * @throws An error if the data could not be loaded.
+   */
+  public async load(): Promise<Document[]> {
```
Comment (bracesproul) [inline langchain/src/document_loaders/web/firecrawl.ts:67]: Prefer `DocumentInterface`
```suggestion
  public async load(): Promise<DocumentInterface[]> {
```

**Review by bracesproul — COMMENTED** [review]
A couple nits, besides that it looks pretty good!

File: docs/core_docs/docs/integrations/document_loaders/web_loaders/firecrawl.mdx (hunk 2 above)
Comment (nickscamara) [reply docs/core_docs/docs/integrations/document_loaders/web_loaders/firecrawl.mdx]: 100%

**Comment by nickscamara** [comment]
@bracesproul Thank you. Just fixed all the suggestions. Let me know if you spot any other ones.

**Review by bracesproul — APPROVED** [review]
lgtm, thanks for adding this!!

**Review by bracesproul — COMMENTED** [review]
hmm, the amount of changed lines in the `yarn.lock` is concerning. I'm going to push up a PR in a second resolving

**Comment by jacoblee93** [comment]
Thank you!

Note that we'll move this to `@langchain/community` in 0.2 in a few weeks.

**Pull request merged**